source/s3_auto_ranged_put.c

```diff
@@ -320,6 +697,40 @@ static int s_s3_auto_ranged_put_prepare_request(
     struct aws_http_message *message = NULL;
 
     switch (request->request_tag) {
+        case AWS_S3_AUTO_RANGED_PUT_REQUEST_TAG_LIST_PARTS: {
+            /* BEGIN CRITICAL SECTION */
+            {
+                aws_s3_meta_request_lock_synced_data(meta_request);
+
+                int message_creation_result = AWS_OP_ERR;
+
+                if (auto_ranged_put->synced_data.list_parts_continuation_token) {
+                    struct aws_byte_cursor continuation_cur =
+                        aws_byte_cursor_from_string(auto_ranged_put->synced_data.list_parts_continuation_token);
+                    message_creation_result = aws_s3_construct_next_request_http_message(
+                        auto_ranged_put->synced_data.list_parts_operation, &continuation_cur, &message);
+                } else {
+                    message_creation_result = aws_s3_construct_next_request_http_message(
+                        auto_ranged_put->synced_data.list_parts_operation, NULL, &message);
+                }
+
+                if (message_creation_result) {
+                    goto message_create_failed;
```

Comment:
add `aws_s3_meta_request_unlock_synced_data(meta_request);` before exist. Otherwise, it could lead to a deadlock.